go/cli/mcap/cmd/du.go

```diff
@@ -0,0 +1,297 @@
+	for {
+		tokenType, data, err := lexer.Next(msg)
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				break
+			}
+
+			return fmt.Errorf("failed to read next token: %w", err)
+		}
+		if len(data) > len(msg) {
+			msg = data
+		}
+
+		instance.totalSize += uint64(len(data))
+		instance.recordKindSize[tokenType.String()] += uint64(len(data))
+
+		switch tokenType {
+		case mcap.TokenChannel:
+			channel, err := mcap.ParseChannel(data)
+			if err != nil {
+				return fmt.Errorf("error parsing Channel: %w", err)
+			}
+
+			instance.channels[channel.ID] = channel
+		case mcap.TokenMessage:
+			message, err := mcap.ParseMessage(data)
+			if err != nil {
+				return fmt.Errorf("error parsing Message: %w", err)
+			}
+			channel := instance.channels[message.ChannelID]
+			if channel == nil {
+				return fmt.Errorf("got a Message record for unknown channel: %d", message.ChannelID)
+			}
+
+			messageSize := uint64(len(message.Data))
+
+			instance.totalMessageSize += messageSize
+			instance.topicMessageSize[channel.Topic] += messageSize
+		case mcap.TokenChunk:
+			chunk, err := mcap.ParseChunk(data)
+			if err != nil {
+				return fmt.Errorf("error parsing Message: %w", err)
+			}
+			err = instance.processChunk(chunk)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	fmt.Println("Top level record stats:")
+	fmt.Println()
+
+	{
+		rows := [][]string{}
+		rows = append(rows, []string{
+			"record",
+			"sum bytes",
+			"% of total file bytes",
+		}, []string{
+			"------",
+			"---------",
+			"---------------------",
+		})
+
+		for recordKind, size := range instance.recordKindSize {
+			row := []string{
+				recordKind, fmt.Sprintf("%d", size),
+				fmt.Sprintf("%f", float32(size)/float32(instance.totalSize)*100.0),
+			}
+
+			rows = append(rows, row)
+		}
+
+		utils.FormatTable(os.Stdout, rows)
+	}
+
+	fmt.Println()
+	fmt.Println("Message size stats:")
+	fmt.Println()
+
+	{
+		rows := [][]string{}
+		rows = append(rows, []string{
+			"topic",
+			"sum bytes (uncompressed)",
+			"% of total message bytes (uncompressed)",
+		}, []string{
+			"-----",
+			"------------------------",
+			"---------------------------------------",
+		})
+
+		type topicInfo struct {
+			name string
+			size uint64
+		}
+		topicInfos := make([]topicInfo, 0, len(instance.topicMessageSize))
+		for topic, size := range instance.topicMessageSize {
+			topicInfos = append(topicInfos, topicInfo{topic, size})
+		}
+
+		// Sort for largest topics first
+		sort.Slice(topicInfos, func(i, j int) bool {
+			return topicInfos[i].size > topicInfos[j].size
+		})
+
+		for _, info := range topicInfos {
+			row := []string{
+				info.name,
+				humanBytes(info.size),
+				fmt.Sprintf("%f", float32(info.size)/float32(instance.totalMessageSize)*100.0),
+			}
+
+			rows = append(rows, row)
+		}
+
+		utils.FormatTable(os.Stdout, rows)
+	}
+
+	return nil
+}
+
+var duCmd = &cobra.Command{
+	Use:   "du <file>",
+	Short: "Report space usage within an MCAP file",
```

Comment:
please also add a `Long` description which includes that notes that this scans the entire file.